File: microsynth/microsynth/report/open_material_requests/open_material_requests.js
// Copyright (c) 2025, Microsynth
// For license information, please see license.txt
/* eslint-disable */


frappe.query_reports["Open Material Requests"] = {
    "filters": [
        {
            fieldname: "supplier",
            label: __("Supplier"),
            fieldtype: "Link",
            options: "Supplier"
        },
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: "Microsynth AG"
        }
    ],
    "onload": (report) => {
        hide_chart_buttons();

        report.page.add_inner_button( __("New Request"), function() {
            // TODO: Open search dialog currently shown on Material Request (Button "Add Item")
        }).addClass("btn-primary");

        if (frappe.user.has_role('Purchase Manager') || frappe.user.has_role('Purchase User')) {
            report.page.add_inner_button( __("Create Purchase Order"), function() {
                create_purchase_order(frappe.query_report.get_filter_values(), frappe.query_report);
            }).addClass("btn-primary");
        }
    },
    'formatter': function(value, row, column, data, default_formatter) {
        // For Item Request rows, show item_name as plain text (no link)
        if (column.fieldname === "item_code" && data.request_type === "Item Request") {
            return data.item_name || value || "";
        }
        return default_formatter(value, row, column, data);
    }
};


function create_purchase_order(filters, report) {
    if (!filters.supplier) {
        frappe.msgprint( __("Please set the Supplier filter"), __("Validation") );
        return;
    }

    // Check for pending Item Requests in report data
    const pendingItemRequests = (report.data || []).filter(row =>
        row.request_type === "Item Request" &&
        row.supplier === filters.supplier
    );

    if (pendingItemRequests.length > 0) {
        frappe.msgprint(__("There are {0} pending Item Requests for this Supplier. Please treat them first.", [pendingItemRequests.length]), __("Warning"));
        return;
    }

    // No pending Item Requests, proceed with PO creation
    frappe.call({
        'method': "microsynth.microsynth.purchasing.create_po_from_open_mr",
        'args': {
            'filters': filters
        },
        'callback': function(r) {
            if (r.message) {
                frappe.set_route("Form", "Purchase Order", r.message);
            } else {
                frappe.show_alert("Internal Error");
            }
        }
    });
}

File: microsynth/microsynth/report/open_material_requests/open_material_requests.py
# Copyright (c) 2025, Microsynth
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _


def get_columns():
    return [
        {"label": _("Material Request"), "fieldname": "material_request", "fieldtype": "Link", "options": "Material Request", "width": 115},
        {"label": _("Request Date"), "fieldname": "transaction_date", "fieldtype": "Date", "width": 95},
        {"label": _("Required By"), "fieldname": "schedule_date", "fieldtype": "Date", "width": 85},
        {"label": _("Item"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 240},
        #{"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Data", "width": 200},
        #{"label": _("Description"), "fieldname": "description", "fieldtype": "Data", "width": 200},
        {"label": _("Qty"), "fieldname": "qty", "fieldtype": "Int", "width": 50},
        {"label": _("Supplier"), "fieldname": "supplier", "fieldtype": "Link", "options": "Supplier", "width": 65},
        {"label": _("Supplier Name"), "fieldname": "supplier_name", "fieldtype": "Data", "width": 200},
        {"label": _("Supplier Part No."), "fieldname": "supplier_part_no", "fieldtype": "Data", "width": 120},
        {"label": _("Requested By"), "fieldname": "requested_by", "fieldtype": "Link", "options": "User", "width": 200}
    ]


def get_data(filters):
    conditions = ""
    if filters and filters.get("supplier"):
        conditions += " AND `tabItem Supplier`.`supplier` = %(supplier)s"

    return frappe.db.sql(f"""
        SELECT
            `tabMaterial Request`.`name` AS material_request,
            `tabMaterial Request`.`transaction_date`,
            `tabMaterial Request Item`.`schedule_date`,
            `tabMaterial Request Item`.`item_code`,
            `tabMaterial Request Item`.`item_name`,
            `tabMaterial Request Item`.`description`,
            `tabMaterial Request Item`.`qty`,
            `tabMaterial Request Item`.`name` AS `material_request_item`,
            `tabItem Supplier`.`supplier`,
            `tabSupplier`.`supplier_name`,
            `tabItem Supplier`.`supplier_part_no`,
            IFNULL(`tabMaterial Request`.`requested_by`, `tabMaterial Request`.`owner`) AS `requested_by`
        FROM
            `tabMaterial Request Item`
        LEFT JOIN `tabMaterial Request`
            ON `tabMaterial Request Item`.`parent` = `tabMaterial Request`.`name`
        LEFT JOIN `tabItem Supplier`
            ON `tabItem Supplier`.`parent` = `tabMaterial Request Item`.`item_code`
            AND `tabItem Supplier`.`parenttype` = 'Item'
            AND `tabItem Supplier`.`idx` = 1
        LEFT JOIN `tabSupplier`
            ON `tabItem Supplier`.`supplier` = `tabSupplier`.`name`
        WHERE
            `tabMaterial Request`.`material_request_type` = 'Purchase'
            AND `tabMaterial Request`.`docstatus` = 1
            AND `tabMaterial Request`.`status` != 'Stopped'
            AND NOT EXISTS (
                SELECT 1
                FROM `tabPurchase Order Item`
                WHERE `tabPurchase Order Item`.`material_request` = `tabMaterial Request Item`.`parent`
                AND `tabPurchase Order Item`.`material_request_item` = `tabMaterial Request Item`.`name`
            )
            {conditions}
        ORDER BY
            `tabMaterial Request`.`schedule_date` ASC
    """, filters, as_dict=True)


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data
